### DISCUSSION TITLE
How to constraint an Optistack variable to be an element of a list?

### DISCUSSION THREAD

**Discussion by LSchroiff** [casadi/casadi#3238] [Q&A]
Hello everyone,

I have the following problem: 
My optimization problem gets a pointcloud `p` (6 x X numpy Array) as Input. The amount of points X may vary depending on the case. I then use Opti stack to define a variable `s` (6 x N). After implementing the goal function, other variables and everything else, I want to implement the constraints. The constraint in question I want to define is `s ∈ p`. The optimization is supposed to run a maximum of `N`-times.

I imported CasADi as `ca` and Opti stack as `opti`.

I tried defining the constraint as:
```
for k in range(N):
  for j in range(len(p)):
    opti.subject_to(ca.logic_any(ca.logic_all(ca.eq(s[:,k], np.transpose(p)[j,:]))))
```

I noticed that `ca.logic_all()` creates a 1x1 matrix, so logical_any() does not work properly in this case. Maybe you have a better idea how to do it in general?

Also I get the following error message, which implies to me that the` ca.logic_...`- operators are not working using an MX variable:
```
17:23:34  Traceback (most recent call last):
  File "C:/Users/CORE_Student4/AppData/Roaming/FreeCAD/FreeCAD_Macros/Create_OptimizedTrajectory.py", line 151, in <module>
    optimization_results = Trajectory_Optimization.trajectory_optimization(all_poses, end_layer, powder_deposition_rate, material_density, deposition_efficiency, nozzle_diameter)
  File "C:\Users/CORE_Student4/AppData/Roaming/FreeCAD/FreeCAD_Macros\Trajectory_Optimization.py", line 264, in trajectory_optimization
    opti.subject_to(ca.logic_any(ca.logic_all(ca.eq(s[:,k], np.transpose(p[j,:])))))
  File "C:\RoboDK\Python37\Lib\site-packages\casadi-windows-py38-v3.5.5-64bit\casadi\casadi.py", line 22223, in logic_all
    return _casadi.logic_all(*args)
<class 'NotImplementedError'>: Wrong number or type of arguments for overloaded function 'logic_all'.
  Possible prototypes are:
    logic_all(DM)
    logic_all(SX)
  You have: '(MX)'
```

Thanks in advance for your help.

**Comment by Marcel1604** [comment]
Hey, this constraint s ∈ p would be interesting for me as well. Does someone have a suggestion?

**Comment by jgillis** [comment]
This sounds like a mixed-integer problem, for which dedicated solvers exist, see https://github.com/casadi/casadi/wiki/FAQ:-how-to-work-with-discrete-mixed-integer-problems%3F
Mixed integer problems are a fundamentally harder than continuous problems, so you experiment with relaxations (i.e. continuous forms that approach the mixed integer problem and have some useful gradient information, unlike the `logic_all` approach).
You could try a weight vector w (0<= w <=1) of length N, penalize it with a strong L1 term (using a slack formulation), and have $s=w.p$.
